spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -199,256 +183,314 @@ else if (messageListener instanceof MessageListener) {
+											&& (elapsed || countExceeded)) {
+										commitIfNecessary();
+										last = now;
+										count = 0;
+									}
+								}
 							}
 						}
 					}
+					else {
+						if (logger.isDebugEnabled()) {
+							logger.debug("No records");
+						}
+					}
 				}
-				else {
-					if (logger.isDebugEnabled()) {
-						logger.debug("No records");
+				catch (WakeupException e) {
+					;
+				}
+				catch (Exception e) {
+					if (getErrorHandler() != null) {
+						getErrorHandler().handle(e, null);
+					}
+					else {
+						logger.error("Container exception", e);
 					}
 				}
 			}
+			if (offsets.size() > 0) {
+				commitIfNecessary();
+			}
+			try {
+				this.consumer.unsubscribe();
+			}
 			catch (WakeupException e) {
 				;
 			}
-			catch (Exception e) {
-				if (getErrorHandler() != null) {
-					getErrorHandler().handle(e, null);
-				}
-				else {
-					logger.error("Container exception", e);
-				}
+			this.consumer.close();
+			if (logger.isInfoEnabled()) {
+				logger.info("Consumer stopped");
 			}
 		}
-		if (this.offsets.size() > 0) {
-			commitIfNecessary();
-		}
-		try {
-			this.consumer.unsubscribe();
-		}
-		catch (WakeupException e) {
-			;
-		}
-		this.consumer.close();
-		if (logger.isInfoEnabled()) {
-			logger.info("Consumer stopped");
-		}
-	}
 
-	private void invokeListener(final ConsumerRecord<K, V> record) {
-		try {
-			if (this.acknowledgingMessageListener != null) {
-				this.acknowledgingMessageListener.onMessage(record, new Acknowledgment() {
+		private void invokeListener(final ConsumerRecord<K, V> record) {
+			try {
+				if (this.acknowledgingMessageListener != null) {
+					this.acknowledgingMessageListener.onMessage(record, new Acknowledgment() {
 
-					@Override
-					public void acknowledge() {
-						if (getAckMode().equals(AckMode.MANUAL)) {
-							updateManualOffset(record);
-						}
-						else if (getAckMode().equals(AckMode.MANUAL_IMMEDIATE)) {
-							updateManualOffset(record);
-							consumer.wakeup();
-						}
-						else {
-							throw new IllegalStateException("AckMode must be MANUAL or MANUAL_IMMEDIATE "
-									+ "for manual acks");
+						@Override
+						public void acknowledge() {
+							if (getAckMode().equals(AckMode.MANUAL)) {
+								updateManualOffset(record);
+							}
+							else if (getAckMode().equals(AckMode.MANUAL_IMMEDIATE)) {
+								if (Thread.currentThread().equals(consumerThread)) {
+									Map<TopicPartition, OffsetAndMetadata> commits = Collections.singletonMap(
+											new TopicPartition(record.topic(), record.partition()),
+											new OffsetAndMetadata(record.offset() + 1));
+									if (logger.isDebugEnabled()) {
+										logger.debug("Committing: " + commits);
+									}
+									consumer.commitAsync(commits, callback);
+								}
+								else {
+									throw new IllegalStateException(
+											"With MANUAL_IMMEDIATE ack mode, acknowledget must be invoked on the "
+											+ "consumer thread");
```

Comment:
Would you mind to elaborate more. Why is it so important?
All other cases uses the same  `commitAsync()`.
With this your change I see no difference between `MANUAL_IMMEDIATE` and `RECORD`.

Thank you!